File: internal/httpapi/blobs.go
package httpapi

import (
	"errors"
	"io"
	"log/slog"
	"net/http"
	"strconv"
	"strings"

	"github.com/bees-roadhouse/hive-sandbox/internal/blob"
	"github.com/bees-roadhouse/hive-sandbox/internal/store"
)

// maxProxyChunk bounds a single proxied response body. A harness run pulling a
// large object gets it in ranges rather than one unbounded stream, so one read
// cannot pin an arbitrary amount of daemon memory or hold a connection open
// indefinitely.
const maxProxyChunk int64 = 32 << 20

// blobRead serves bytes to a caller that already holds a reference to them.
//
// This is the same authorization the guest capability performs, over HTTP
// instead of the ABI, and deliberately the SAME resolution path: reads go
// through the caller's refs, never the global hash space. A harness container
// reaches this over the bind-mounted unix socket with --network=none, so it can
// read what it was given without an IP network existing at all.
//
// What this endpoint must never become is a way to turn a hash into bytes.
// "No such blob" and "a blob exists and you hold no ref to it" are the same
// answer here for the same reason they are in GuestBlobs: distinguishing them
// makes a content address an existence oracle.
func (a *API) blobRead(w http.ResponseWriter, r *http.Request, cred store.Credential) {
	ctx := r.Context()

	h, err := blob.ParseHash(r.PathValue("hash"))
	if err != nil {
		fail(w, http.StatusBadRequest, "malformed blob address")
		return
	}

	rng, err := parseRange(r.Header.Get("Range"))
	if err != nil {
		fail(w, http.StatusRequestedRangeNotSatisfiable, "bad range")
		return
	}

	desc, level, rc, err := a.blobs.Open(ctx, cred, h, rng)
	if err != nil {
		// One shape for every failure to produce bytes, and a log line carrying
		// the detail so an operator can still tell the cases apart.
		slog.Warn("blob read refused",
			"actor", cred.ActorID, "principal", cred.PrincipalID,
			"blob", h.String(), "err", err)
		fail(w, http.StatusNotFound, "blob not found")
		return
	}
	defer func() { _ = rc.Close() }()

	w.Header().Set("Content-Type", desc.MIME)
	w.Header().Set("Accept-Ranges", "bytes")
	// Provenance travels with the bytes. A caller that stores or forwards them
	// has been told what they are, so nothing downstream has to guess -- and an
	// untrusted blob stays untrusted across an HTTP hop the way it does across
	// the ABI (invariant 9).
	w.Header().Set("X-Hive-Trust", string(level.Normalize()))
	w.Header().Set("X-Hive-Blob", h.String())

	status := http.StatusOK
	if !rng.IsFull() {
		status = http.StatusPartialContent
	}
	w.WriteHeader(status)

	if r.Method == http.MethodHead {
		return
	}
	// A write failure here means the client went away mid-body. Nothing to do
	// about it and nothing worth an error line.
	_, _ = io.Copy(w, io.LimitReader(rc, maxProxyChunk))
}

// parseRange understands the single-range form, which is what a shell client
// and every HTTP library actually send.
//
// Multi-range responses are multipart/byteranges, and a caller that wanted two
// windows can ask twice. Accepting the syntax and serving one range would be
// worse than refusing it: the client would believe it had both.
func parseRange(header string) (blob.Range, error) {
	header = strings.TrimSpace(header)
	if header == "" {
		return blob.Range{}, nil
	}
	spec, ok := strings.CutPrefix(header, "bytes=")
	if !ok || strings.Contains(spec, ",") {
		return blob.Range{}, errors.New("unsupported range")
	}
	start, end, ok := strings.Cut(spec, "-")
	if !ok {
		return blob.Range{}, errors.New("malformed range")
	}
	// A suffix range ("-500", the last 500 bytes) needs the object size to
	// resolve, which the driver has and this parser does not.
	if strings.TrimSpace(start) == "" {
		return blob.Range{}, errors.New("suffix ranges are not supported")
	}
	offset, err := strconv.ParseInt(strings.TrimSpace(start), 10, 64)
	if err != nil || offset < 0 {
		return blob.Range{}, errors.New("bad range start")
	}
	if strings.TrimSpace(end) == "" {
		// "bytes=N-" is from N to the end, which a zero Length already means.
		return blob.Range{Offset: offset}, nil
	}
	last, err := strconv.ParseInt(strings.TrimSpace(end), 10, 64)
	if err != nil || last < offset {
		return blob.Range{}, errors.New("bad range end")
	}
	// HTTP ranges are inclusive of the last byte; blob.Range is a length.
	return blob.Range{Offset: offset, Length: last - offset + 1}, nil
}

File: crates/hive-httpapi/src/blobs.rs
//! Blob reads over HTTP: the same authorization the guest capability performs.

use axum::body::Body;
use axum::extract::{Path, State};
use axum::response::{IntoResponse, Response};
use hive_blob::{Hash, Range};
use hive_httpauth::Authed;
use http::{HeaderMap, Method, StatusCode, header};
use tokio::io::AsyncReadExt;

use crate::{AppState, fail};

/// Bounds a single proxied response body. A harness run pulling a large object
/// gets it in ranges rather than one unbounded stream.
const MAX_PROXY_CHUNK: u64 = 32 << 20;

/// Serves bytes to a caller that already holds a reference to them. Reads go
/// through the caller's refs, never the global hash space. "No such blob" and
/// "a blob exists and you hold no ref to it" are the same answer here for the
/// same reason they are in the guest capability: distinguishing them makes a
/// content address an existence oracle.
pub(crate) async fn read(
    State(s): State<AppState>,
    Authed(cred): Authed,
    method: Method,
    Path(hash): Path<String>,
    headers: HeaderMap,
) -> Response {
    let Some(blobs) = &s.blobs else {
        return fail(StatusCode::NOT_FOUND, "not found");
    };
    let Ok(h) = Hash::parse(&hash) else {
        return fail(StatusCode::BAD_REQUEST, "malformed blob address");
    };
    let range_header = headers
        .get(header::RANGE)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");
    let Ok(rng) = parse_range(range_header) else {
        return fail(StatusCode::RANGE_NOT_SATISFIABLE, "bad range");
    };
    let (desc, level, rc) = match blobs.open(&cred, h, rng).await {
        Ok(x) => x,
        Err(e) => {
            tracing::warn!(actor = %cred.actor_id, principal = %cred.principal_id, blob = %h, err = %e, "blob read refused");
            return fail(StatusCode::NOT_FOUND, "blob not found");
        }
    };
    let status = if rng.is_full() {
        StatusCode::OK
    } else {
        StatusCode::PARTIAL_CONTENT
    };
    let mut resp = Response::builder()
        .status(status)
        .header(header::CONTENT_TYPE, desc.mime.as_str())
        .header(header::ACCEPT_RANGES, "bytes")
        // Provenance travels with the bytes: an untrusted blob stays
        // untrusted across an HTTP hop the way it does across the ABI.
        .header("x-hive-trust", level.as_str())
        .header("x-hive-blob", h.to_string());
    if method == Method::HEAD {
        return resp.body(Body::empty()).expect("static response");
    }
    let stream = tokio_util::io::ReaderStream::new(rc.take(MAX_PROXY_CHUNK));
    resp = resp.header("x-hive-size", desc.size.to_string());
    resp.body(Body::from_stream(stream))
        .unwrap_or_else(|_| fail(StatusCode::INTERNAL_SERVER_ERROR, "internal").into_response())
}

/// The single-range form, which is what a shell client and every HTTP library
/// actually send. Multi-range responses are multipart/byteranges, and a caller
/// that wanted two windows can ask twice; accepting the syntax and serving one
/// range would be worse than refusing it.
pub fn parse_range(header: &str) -> Result<Range, &'static str> {
    let header = header.trim();
    if header.is_empty() {
        return Ok(Range::FULL);
    }
    let spec = header.strip_prefix("bytes=").ok_or("unsupported range")?;
    if spec.contains(',') {
        return Err("unsupported range");
    }
    let (start, end) = spec.split_once('-').ok_or("malformed range")?;
    // A suffix range ("-500") needs the object size, which the driver has and
    // this parser does not.
    if start.trim().is_empty() {
        return Err("suffix ranges are not supported");
    }
    let offset: u64 = start.trim().parse().map_err(|_| "bad range start")?;
    if end.trim().is_empty() {
        return Ok(Range::new(offset, 0));
    }
    let last: u64 = end.trim().parse().map_err(|_| "bad range end")?;
    if last < offset {
        return Err("bad range end");
    }
    // HTTP ranges are inclusive of the last byte; a Range is a length.
    Ok(Range::new(offset, last - offset + 1))
}

#[cfg(test)]
mod tests {
    use super::*;

    /// "bytes=0-0" is one byte, not zero.
    #[test]
    fn parse_range_converts_inclusive_end_to_length() {
        for (header, offset, length) in [
            ("", 0, 0),
            ("bytes=0-0", 0, 1),
            ("bytes=0-99", 0, 100),
            ("bytes=100-199", 100, 100),
            ("bytes=500-", 500, 0),
            ("  bytes=0-9  ", 0, 10),
        ] {
            let got =
                parse_range(header).unwrap_or_else(|e| panic!("parse_range({header:?}) = {e}"));
            assert_eq!((got.offset, got.length), (offset, length), "{header:?}");
        }
    }

    /// A range we cannot serve correctly must be REFUSED, never narrowed.
    #[test]
    fn parse_range_refuses_what_it_cannot_serve() {
        for header in [
            "bytes=0-99,200-299",
            "bytes=-500",
            "bytes=99-0",
            "bytes=-",
            "bytes=abc-def",
            "items=0-99",
            "bytes=0-99extra",
            "bytes=-1-5",
        ] {
            assert!(
                parse_range(header).is_err(),
                "parse_range({header:?}) accepted"
            );
        }
    }
}
